### PR TITLE
Add rsync daemon copy strategy for 'oc rsync' windows support

### PR DESCRIPTION
**Pull request by csrwng**
Reworks the rsync command to use one of 3 strategies when copying: rsync, rsync-daemon or tar. The extended test now tests all 3 strategies.

Tar support in windows depends on https://github.com/openshift/source-to-image/pull/316.

### CODE REVIEW

**Comment by csrwng** [comment]
@bparees @smarterclayton ptal

File: pkg/cmd/cli/cmd/rsync/local_exec.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package rsync
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/rsync/local_exec.go]: No space in the filename

File: pkg/cmd/cli/cmd/rsync/remote_exec.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
+package rsync
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/rsync/remote_exec.go]: If these are different strategies, give each file a common prefix like `exec_remote.go`, `exec_local.go`, `exec_forward.go`

File: pkg/cmd/cli/cmd/rsync/rsync.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+)
+
+const (
+	// RsyncRecommendedName is the recommended name for the rsync command
+	RsyncRecommendedName = "rsync"
+
+	rsyncLong = `
+Copy local files to a container
+
+This command will attempt to copy local files to a remote container. It will default to the
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/rsync/rsync.go]: I don't know how useful the details of the algorithm are.  Instead, I'd use this space to explain what it does (copy files that have changed locally to the remote system) and add instructions about the different operating systems (on Linux and Mac, install rsync.  On Windows, download cwRsync)

File: pkg/cmd/cli/cmd/rsync/rsync.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+			kcmdutil.CheckErr(o.Complete(f, c, args))
+			kcmdutil.CheckErr(o.Validate())
+			kcmdutil.CheckErr(o.RunRsync())
+		},
+	}
+
+	cmd.Flags().StringVarP(&o.ContainerName, "container", "c", "", "Container within the pod")
+	cmd.Flags().BoolVarP(&o.Quiet, "quiet", "q", false, "Quiet copy")
+	cmd.Flags().BoolVar(&o.Delete, "delete", false, "Delete files not present in source")
+	cmd.Flags().StringVar(&o.StrategyName, "strategy", "", "Specify which strategy to use for copy: rsync, rsync-daemon, or tar")
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/rsync/rsync.go]: Move this flag with container name

File: pkg/cmd/cli/cmd/rsync/rsync.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+		Example: fmt.Sprintf(rsyncExample, parent+" "+name),
+		Run: func(c *cobra.Command, args []string) {
+			kcmdutil.CheckErr(o.Complete(f, c, args))
+			kcmdutil.CheckErr(o.Validate())
+			kcmdutil.CheckErr(o.RunRsync())
+		},
+	}
+
+	cmd.Flags().StringVarP(&o.ContainerName, "container", "c", "", "Container within the pod")
+	cmd.Flags().BoolVarP(&o.Quiet, "quiet", "q", false, "Quiet copy")
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/rsync/rsync.go]: Put these two flags separated from other flags and add a comment "These flags must match rsync flag names"

File: pkg/cmd/cli/cmd/rsync/pathspec.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+	return &pathSpec{
+		PodName: parts[0],
+		Path:    parts[1],
+	}, nil
+}
+
+// convertWindowsPath converts a windows-specific path to a path that can be used by
+// the rsync command in windows
+func convertWindowsPath(path string) string {
+	parts := strings.SplitN(path, ":", 2)
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/rsync/pathspec.go]: Hrm.  Yuck.

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/rsync/pathspec.go]: Can you add more doc explaining this?

Comment (csrwng) [inline pkg/cmd/cli/cmd/rsync/pathspec.go]: sure... basically windows rsync requires paths to be specified as cygwin paths... I didn't think it'd be acceptable to require our users to do that.

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/rsync/pathspec.go]: I just want the comment in the code so someone knows why when they're
debugging

On Oct 23, 2015, at 3:51 PM, Cesar Wong <notifications@github.com> wrote:

In pkg/cmd/cli/cmd/rsync/pathspec.go
<https://github.com/openshift/origin/pull/5351#discussion_r42907192>:

> +		}, nil
> +	}
> +	valid, msg := kvalidation.ValidatePodName(parts[0], false)
> +	if !valid {
> +		return nil, fmt.Errorf("invalid pod name %s: %s", parts[0], msg)
> +	}
> +	return &pathSpec{
> +		PodName: parts[0],
> +		Path:    parts[1],
> +	}, nil
> +}
> +
> +// convertWindowsPath converts a windows-specific path to a path that can be used by
> +// the rsync command in windows
> +func convertWindowsPath(path string) string {
> +	parts := strings.SplitN(path, ":", 2)

sure... basically windows rsync requires paths to be specified as cygwin
paths... I didn't think it'd be acceptable to require our users to do that.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5351/files#r42907192>.

**Comment by csrwng** [comment]
Addressed comments so far 
[test]

**Comment by csrwng** [comment]
[test]

**Comment by smarterclayton** [comment]
[extended(core)]

**Comment by smarterclayton** [comment]
[testonlyextended][extended:core]

**Comment by smarterclayton** [comment]
LGTM [merge] so we have more soak time.

**Comment by csrwng** [comment]
Looks like a flake, retrying the [merge]

**Pull request merged**